### PR TITLE
Quarantine media by ID or user ID

### PR DESCRIPTION
**Pull request by anoadragon453**
Fixes https://github.com/matrix-org/synapse/issues/5956

This PR adds some admin APIs to:

* Quarantine media by ID
* Quarantine all local media that a local user has uploaded

Most of the code here is test-related.

### CODE REVIEW

File: synapse/storage/data_stores/main/room.py [hunk 1]
```diff
@@ -938,6 +938,218 @@ def block_room(self, room_id, user_id):
...
+
+        def _quarantine_media_in_room_txn(txn):
+            local_mxcs, remote_mxcs = self._get_media_mxcs_in_room_txn(txn, room_id)
+            return self._quarantine_media_txn(
+                txn, local_mxcs, remote_mxcs, quarantined_by,
+            )
+
+        return self.db.runInteraction(
+            "quarantine_media_in_room", _quarantine_media_in_room_txn
+        )
```
Comment (erikjohnston) [inline synapse/storage/data_stores/main/room.py]: These functions got moved up to the `RoomWorkerStore`, so you probably want to delete the duplicate functions and move the new ones up to the worker store too.

File: tests/rest/admin/test_admin.py [hunk 2]
```diff
@@ -346,3 +352,326 @@ def test_purge_room(self):
...
+        hs = self.setup_test_homeserver(config=config, http_client=client)
+
+        return hs
+
+    def test_quarantine_media_requires_admin(self):
+        self.register_user("nonadmin", "pass", admin=False)
+        non_admin_user_tok = self.login("nonadmin", "pass")
+
+        # Attempt quarantine media APIs as non-admin
+        url = "/_synapse/admin/v1/quarantine_media/id/example.org/abcde12345"
```
Comment (erikjohnston) [inline tests/rest/admin/test_admin.py]:
```suggestion
        url = "/_synapse/admin/v1/quarantine_media/example.org/abcde12345"
```

File: synapse/storage/data_stores/main/room.py [hunk 3]
```diff
@@ -938,6 +938,218 @@ def block_room(self, room_id, user_id):
...
+            """
+        if filter_quarantined:
+            sql += "AND quarantined_by IS NULL"
+        txn.execute(sql, (user_id,))
+
+        rows = txn.fetchall()
+        if rows:
+            local_media_ids = [row[0] for row in rows]
+        else:
+            local_media_ids = []
```
Comment (erikjohnston) [inline synapse/storage/data_stores/main/room.py]: Its more efficient to replace all this with:

```python
local_media_ids = [media_id for media_id, in txn]
```

As this skips creating (a potentially large) list that we just throw away.

File: tests/rest/admin/test_admin.py (hunk 2 above)
Comment (anoadragon453) [reply tests/rest/admin/test_admin.py]: I've changed the endpoints to:

- `/_synapse/admin/v1/quarantine_media/by_id/example.org/abcde12345`
- `/_synapse/admin/v1/quarantine_media/by_user/@bob:example.com`
- `/_synapse/admin/v1/quarantine_media/by_room/!someroom:example.com` (with `.../quarantine_media/!room...` as a legacy fallback)

Comment (anoadragon453) [reply tests/rest/admin/test_admin.py]: Another iteration:

* `/_synapse/admin/v1/quarantine_media_by_id/example.org/abcde12345`
* `/_synapse/admin/v1/quarantine_media_by_user/@bob:example.com`
* `/_synapse/admin/v1/quarantine_media_by_room/!someroom:example.com` (with `.../quarantine_media/!room...` as a legacy fallback)

File: docs/admin_api/media_admin_api.md [hunk 4]
```diff
@@ -22,19 +22,55 @@ It returns a JSON body like the following:
...
-The media file itself (and any thumbnails) is not deleted from the server.
+## Quarantining media in a room
+
+This API quarantines all local and remote media in a room.
+
+```
+POST /_synapse/admin/v1/quarantine_media_by_room/<room_id>
+
+{
+  "num_quarantined": 10  # The number of media items successfully quarantined
```
Comment (richvdh) [inline docs/admin_api/media_admin_api.md]: is this a thing you need to pass in the POST body? the doc makes it look that way. likewise below

Comment (anoadragon453) [reply docs/admin_api/media_admin_api.md]: Very good point.

File: synapse/storage/data_stores/main/room.py [hunk 5]
```diff
@@ -938,6 +1052,34 @@ def block_room(self, room_id, user_id):
             (room_id,),
         )
 
+    def get_media_mxcs_in_room(self, room_id):
```
Comment (erikjohnston) [inline synapse/storage/data_stores/main/room.py]: This is still a dupe?

File: changelog.d/6681.feature [hunk 6]
```diff
@@ -0,0 +1 @@
+Extend the quarantine_media admin API to quarantine media by ID or all media by a specific user.
```
Comment (erikjohnston) [inline changelog.d/6681.feature]:
```suggestion
Add new quarantine media admin APIs to quarantine by media ID or by user who uploaded the media.
```

**Comment by anoadragon453** [comment]
Erik and I tried to point these tests at the media worker, however it doesn't have servlets for creating or joining a room (which the test for quarantining media in a room needs), so we need some way of testing against multiple workers.

However, we're not going to block this PR on that.

File: synapse/rest/admin/media.py [hunk 7]
```diff
@@ -32,23 +32,85 @@ class QuarantineMediaInRoom(RestServlet):
     this server.
     """
 
-    PATTERNS = historical_admin_path_patterns("/quarantine_media/(?P<room_id>[^/]+)")
+    PATTERNS = (
+        historical_admin_path_patterns("/room/(?P<room_id>[^/]+)/media/quarantine")
+        +
+        # This path kept around for legacy reasons
+        historical_admin_path_patterns("/quarantine_media/(?P<room_id>![^/]+)")
```
Comment (anoadragon453) [inline synapse/rest/admin/media.py]: Extra `!` snuck in here :man_facepalming:

**Pull request merged**